statsmodels/regression/special_linear_model.py

```diff
@@ -0,0 +1,182 @@
+
+    This is intended for almost balanced panels. The data is converted
+    to a 3-dimensional array with nans for missing cells.
+
+    currently works only for two-way effects
+    groups have to be integers corresponding to range(k_cati)
+
+    no input error checking
+
+    This function will change as more options and special cases are
+    included.
+
+    Parameters
+    ----------
+    exog_dense : 2d ndarray
+        data with observations in rows and variables in columns.
+        This array will currently not be modified.
+    groups : 2d ndarray, int
+        groups labels specified as consecutive integers starting at zero
+    add_mean : bool
+        If true (default), then the total variable means are added back into
+        the group demeand exog_dense
+    max_iter : int
+        maximum number of iterations
+    atol : float
+        tolerance for convergence. Convergence is achieved if the
+        maximum absolute change (np.ptp) is smaller than atol.
+
+    Returns
+    -------
+    ex_dm_w : ndarray
+        group demeaned exog_dense array in wide format
+    ex_dm : ndarray
+        group demeaned exog_dense array in long format
+    it : int
+        number of iterations used. If convergence has not been
+        achieved then it will be equal to max_iter - 1
+
+    """
+
+    # with unbalanced panel
+
+    k_cat = tuple((groups.max(0) + 1).tolist())
+    xm = np.empty(exog_dense.shape[1:] + k_cat)
+    xm.fill(np.nan)
+    xm[:, groups[:, 0], groups[:, 1]] = exog_dense.T
+    keep = ~np.isnan(xm[0]).ravel()
+    finished = False
+    for it in range(max_iter):
+        for axis in range(1, xm.ndim):
+            group_mean = np.nanmean(xm, axis=axis, keepdims=True)
+            xm -= group_mean
+            if np.ptp(group_mean) < atol:
+                finished = True
+                break
+        if finished:
+            break
+
+    xd = xm.reshape(exog_dense.shape[-1], -1).T[keep]
+    if add_mean:
+        xmean = exog_dense.mean(0)
+        xd += xmean
+        xm += xmean[:, None, None]
+    return xm, xd, it
+
+
+class OLSAbsorb(WLS):
+    """OLS model that absorbs categorical explanatory variables
+
+    see docstring for OLS, the following only has the extra parameters for now
+
+    Parameters
+    ----------
+    exog_absorb : ndarray, 1D or 2D
+        categorical, factor variables that will be absorbed
+    absorb_method : string, 'lu' or 'lmgres'
+        method used in projection for absorbing the factor variables.
+        Currently the options use either sparse LU decomposition or sparse
+        `lmgres`
+
+    Notes
+    -----
+    constant: the current parameterization produces a constant when the mean
+    categorical effect is set to zero
+
+    Warning: currently not all inherited methods for OLS are correct.
+    Parameters and inference are correct and correspond to the full model for
+    OLS that includes all factor variables as dummies with zero mean factor
+    effects.
```

Comment:
add comment about `predict`, ... that differ from OLS, at least in this version
also untested robust cov_params, differences in HC2, HC3